File: LeetCode_Java_ＷinterBreak2014/deleteDuplicates.java
import com.ListNode;

import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;

public class deleteDuplicates {
    public static class Solution {
        public ListNode deleteDuplicates0(ListNode head) {
            if (head == null || head.next == null) return head;  // forgot this line
            
            ListNode res = head;
            ListNode curr = head.next;
            ListNode prev = head;
            ListNode fast = curr;
            while (curr != null) {
                if (curr.val == prev.val) {
                    fast = curr.next;
                    prev.next = fast;
                    curr.next = null;
                    curr = prev.next;
                } else {
                    prev = curr;
                    curr = curr.next;
                }
            }
            return res;
        }

        public void recursion(ListNode prev, ListNode curr) {
            if (curr == null) return;
            if (prev.val == curr.val) {
                prev.next = curr.next;
                recursion(prev, curr.next);
            } else {
                prev.next = curr;   // important for Java
                recursion(prev.next, curr.next);
            }
        }
        
        public ListNode deleteDuplicates1(ListNode head) {
            if (head == null) return head;
            ListNode dummy = new ListNode(head.val + 1);
            recursion(dummy, head);
            return dummy.next;
        }

        public ListNode deleteDuplicates(ListNode head) {
            if (head == null) return head;
            for (ListNode prev = head, curr = head.next; curr != null; curr = curr.next) {
                if (prev.val == curr.val) {
                    prev.next = curr.next;
                } else {
                    prev = curr;
                }
            }
            return head;
        }
    }

    public static void main(String[] args){
        Solution result = new Solution();
        int [] a = {1, 1};
        ListNode head = new ListNode(1);
        head.buildList(head, a);
        head.printList(head);
        ListNode res = result.deleteDuplicates(head);
        System.out.println("");
        while (res != null) {
            System.out.print(res.val + ", ");
            res = res.next;
        }
    }
}

File: myalgorithms/LeetCode_c++_0325-0430_2014-100/deleteDuplicates.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};

ListNode *deleteDuplicates(ListNode *head) {
    if (head == NULL) return NULL;

    ListNode *prev = head;
    ListNode *curr = head->next;
    unordered_map<int, bool> hash;
    hash[head->val] = true;

    while(curr) {
        if (hash[curr->val]) {
            ListNode* tmp = curr;
            curr = curr->next;
            prev->next = curr;
            delete tmp;
        } else {
            hash[curr->val] = true;
            prev = curr;
            curr = curr->next;
        }

    }
    return head;
}

int main() {
    ListNode* prev;
    ListNode* curr;

    int n = 7;
    int a[] = {-1, 0, 0, 0, 0, 3, 3};
        
    for (int i = 0; i < n; ++i) {
        if (i == 0) {
            prev = new ListNode(a[i]);
            curr = prev;
        } else  {
            curr->next = new ListNode(a[i]);
            curr = curr->next;
        }
    }

    curr = prev;
    while(curr)  {
        cout << curr->val << ", " << endl;
        curr = curr->next;
    }
    
    deleteDuplicates(prev);
    while(prev)  {
        cout << prev->val << ", " << endl;
        prev = prev->next;
    }


    return 0;
}
